server/ctrlsubsonic/handlers_by_tags.go

```diff
@@ -380,18 +412,84 @@ func (c *Controller) ServeGetSongsByGenre(r *http.Request) *spec.Response {
 		List: make([]*spec.TrackChild, len(tracks)),
 	}
 	for i, track := range tracks {
-		sub.TracksByGenre.List[i] = spec.NewTrackByTags(track, track.Album)
+		tr := spec.NewTrackByTags(track, track.Album)
+		sub.TracksByGenre.List[i] = tr
 	}
 	return sub
 }
 
 func (c *Controller) ServeGetStarredTwo(r *http.Request) *spec.Response {
+	user := r.Context().Value(CtxUser).(*db.User)
+	params := r.Context().Value(CtxParams).(params.Params)
+	m := c.getMusicFolder(params)
+
 	sub := spec.NewResponse()
 	sub.StarredTwo = &spec.StarredTwo{
 		Artists: []*spec.Artist{},
 		Albums:  []*spec.Album{},
 		Tracks:  []*spec.TrackChild{},
 	}
+
+	// artists
+	var artists []*db.Artist
+	q := c.DB.Table("artists").Joins("JOIN artist_stars on artists.id=artist_stars.artist_id").
+		Where("artist_stars.user_id=?", user.ID).
+		Preload("ArtistStar", "user_id=?", user.ID).
+		Preload("ArtistRating", "user_id=?", user.ID)
+
+	if m != "" {
+		q = q.Joins("JOIN albums on artists.id=albums.tag_artist_id").Where("albums.rootdir=?", m)
+	}
+	if err := q.Find(&artists).Error; err != nil {
+		return spec.NewError(0, "find artists: %v", err)
+	}
+	for _, a := range artists {
+		artist := spec.NewArtistByTags(a)
+		sub.StarredTwo.Artists = append(sub.StarredTwo.Artists, artist)
+	}
+
+	// albums
+	var albums []*db.Album
+	q = c.DB.Table("albums").Joins("JOIN album_stars on albums.id=album_stars.album_id").
+		Where("album_stars.user_id=?", user.ID).
+		Preload("AlbumStar", "user_id=?", user.ID).
+		Preload("AlbumRating", "user_id=?", user.ID)
+	if m != "" {
+		q = q.Where("rootdir=?", m)
+	}
+	if err := q.Find(&albums).Error; err != nil {
+		return spec.NewError(0, "find albums: %v", err)
+	}
+	for _, a := range albums {
+		var ar db.Artist
+		if err := c.DB.Where("artist_id=?", a.TagArtistID).Find(&ar); err != nil {
+			return spec.NewError(0, "find artist for album: %v", err)
+		}
+		album := spec.NewAlbumByTags(a, &ar)
+		sub.StarredTwo.Albums = append(sub.StarredTwo.Albums, album)
+	}
+
+	// tracks
+	var tracks []*db.Track
+	q = c.DB.Table("tracks").Joins("right join track_stars on tracks.id=track_stars.track_id").
```

Comment:
the right joins are giving me an error like 

`subsonic error code 0: find tracks: RIGHT and FULL OUTER JOINs are not currently supported`

in this case it seems like a normal inner join would do the trick? or wdyt?